fix Python 3 string/map issues

pdebuyl opened the pull request:
Description of changes:
1. explicitly set encoding in set_random_state_PRNG in
   src/python/espressomd/system.pyx
2. use (c)StringIO from six in src/python/espressomd/MDA_ESP/__init__.py
3. "listify" a map in src/python/espressomd/MDA_ESP/__init__.py

There were a few Python 3 incompatibilities I encountered.

I used https://stackoverflow.com/questions/42678779/how-to-pass-string-from-python3-to-cythonized-c-function for the Python 3 `libcpp.string` vector issue.

PR Checklist
------------
 - [ ] Tests?
   - [ ] Interface
   - [ ] Core 
 - [ ] Docs?

pdebuyl committed:
fix Python 3 string/map issues

1. explicitly set encoding in set_random_state_PRNG in
   src/python/espressomd/system.pyx
2. use (c)StringIO from six in src/python/espressomd/MDA_ESP/__init__.py
3. "listify" a map in src/python/espressomd/MDA_ESP/__init__.py

hmenke commented on src/python/espressomd/MDA_ESP/__init__.py:
```diff
@@ -30,7 +30,7 @@
 
 """
-import cStringIO
+from six import StringIO
```
`six` is not part of the Python standard library.  I would prefer if the code wasn't introducing additional dependencies.

pdebuyl replied:
There is already an import of six in `src/python/espressomd/io/writer/h5md.py` and MDAnalysis requires it anyway. If you prefer, I can replace this by a try/except block or a conditional import that check the version of Python.

pdebuyl replied:
Also, if needed, [six](https://github.com/benjaminp/six)

> (...) is contained in only one Python file, so it can be easily copied into your project.

But it is on PyPI and does not carry further dependencies.

mkuron replied:
We don't need to use six for this. What it does is trivial and you can do the same using something like:
```
try:
    import cStringIO as StringIO
except ImportError:
    from io import StringIO
```

hmenke commented on src/python/espressomd/system.pyx:
```diff
@@ -330,7 +330,7 @@ cdef class System(object):
                 states_on_node_i.append(
                     rng.randint(0, numeric_limits[int].max()))
-            states[i] = " ".join(map(str, states_on_node_i))
+            states[i] = (" ".join(map(str, states_on_node_i))).encode('utf-8')
```
Wouldn't it be easier to use a Unicode literal string? (untested)
```python
states[i] = u" ".join(map(str, states_on_node_i))
```

pdebuyl replied:
It does not work with either of `u` or `b` as a prefix. As we store string in a `std::vector[std::string]` at the Cython level for Python 2/3 compatibility, I did proceed by trial and error and then found the SO answer I mention in the PR.

pdebuyl committed:
fix Python 3 string/map issues

1. explicitly set encoding for the of the property property
   random_number_generator_state.
2. "listify" the map for the return value of random_number_generator_state.
   This change simplifies the calling code, such as in
   samples/minimal_random_number_generator.py

pdebuyl commented:
@hmenke I realized that two similar changes were needed in system.pyx, I pushed them on the same PR, I hope this is not an issue.

pdebuyl commented:
@hmenke is there any extra task needed for this PR on my side?

fweik commented:
I'm not thrilled about the extra dependency. h5md and MDAnalysis are not required dependencies of Espresso so far, so six is not yet required to build or run it.
Can you please explain why the changes you are introducing here are required? As far as I can tell the code was working fine with python 3 before (the tests are currently run on python 3.5.2).

pdebuyl committed:
remove use of six

pdebuyl commented:
Ok, as the PR was already approved I believed this was ok. I made the change as requested.

Note however that anyone actually using the `MDA_ESP` module will require mdanalysis that requires six anyway (and about ten thousands other packages of course).

pdebuyl commented:
@fweik with respect to testing, the issues originated is some of the samples which are not part of the testsuite I suppose. Should I consider a failure to run a test a good reason to submit other PRs?

fweik commented:
Yes, fixes for the samples are also very welcome. They often slip thru because they are not automatically tested at the moment, but of course they should work.
